packages/non-trapping-shim/src/non-trapping-pony.js

```diff
@@ -0,0 +1,289 @@
+     * need to meta-handle that `handlerPlus[trapName]`.
+     *
+     * @param {any} target
+     * @param {any[]} rest
+     */
+    const trapPlus = freeze((target, ...rest) => {
+      if (isNonTrappingInternal(target, true)) {
+        defineProperty(handlerPlus, trapName, {
+          value: undefined,
+          writable: false,
+          enumerable: true,
+          configurable: false,
+        });
+      } else {
+        if (!hasOwn(handlerPlus, trapName)) {
+          defineProperty(handlerPlus, trapName, {
+            value: trapPlus,
+            writable: false,
+            enumerable: true,
+            configurable: true,
+          });
+        }
+        const { [ORIGINAL_HANDLER]: originalHandler } = handlerPlus;
+        const trap = originalHandler[trapName];
+        if (trap !== undefined) {
+          // Note that whether `trap === undefined` can change dynamically,
+          // so we do not install an own `handlerPlus[trapName] === undefined`
+          // for that case. We still install or preserve an own
+          // `handlerPlus[trapName] === trapPlus` until the target is
+          // seen to be non-trapping.
+          return apply(trap, originalHandler, [target, ...rest]);
+        }
+      }
+      return ReflectPlus[trapName](target, ...rest);
+    });
+    return trapPlus;
+  },
+});
+
+/**
+ * A handlerPlus starts as a fresh empty object that inherits from a proxy
+ * whose handler is the shared generic metaHandler.
+ * Thus, the metaHandler's `get` method is called only when the
+ * `handlerPlus` does not have a property overriding that `trapName`.
+ * In that case, the metaHandler's `get` is called with its `receiver`
+ * being the `handlerPlus`.
+ *
+ * @param {ProxyHandler<any>} originalHandler
+ * @returns {ProxyHandler<any> & {
+ *   isNonTrapping: (target: any) => boolean,
+ *   suppressTrapping: (target: any) => boolean,
+ *   originalHandler: ProxyHandler<any>
+ * }}
+ */
+const makeHandlerPlus = originalHandler => ({
+  // @ts-expect-error TS does not know what this __proto__ is doing
+  __proto__: new OriginalProxy({}, metaHandler),
+  [ORIGINAL_HANDLER]: originalHandler,
+});
+
+const ProxyInternal = function Proxy(target, handler) {
+  if (new.target !== ProxyInternal) {
+    if (new.target === undefined) {
+      throw TypeError('Proxy constructor requires "new"');
+    }
+    throw TypeError('Safe Proxy shim does not support subclassing');
+  }
+  const handlerPlus = makeHandlerPlus(handler);
+  const proxy = new OriginalProxy(target, handlerPlus);
+  proxyHandlerMap.set(proxy, [target, handler]);
+  return proxy;
+};
+
+/**
+ * In the shim, `ProxyPlus` replaces the global `Proxy`.
+ *
+ * We use `bind` as the only way for user code to produce a
+ * constructible function (i.e., one that responds to `new`) without a
+ * `.prototype` property.
+ *
+ * @type {ProxyConstructor}
+ */
+const ProxyPlus = ProxyInternal.bind(undefined);
```

Comment:
You do need to redefine `.name` if you want to keep the original `"Proxy"` name instead `"bound Proxy"`

Reply:
Done